### PR TITLE
[image_picker] Added maxDuration property to pickVideo method

### PR DESCRIPTION
**Pull request by derTuca**
## Description

Adds a `maxDuration` property so a maximum recording duration can be set.

## Related Issues

flutter/flutter#40035

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by qwerqwq** [comment]
image_picker有压缩图片的api吗

File: packages/image_picker/ios/Classes/ImagePickerPlugin.m [hunk 1]
```diff
@@ -93,6 +93,11 @@ - (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result
     _arguments = call.arguments;
 
     int imageSource = [[_arguments objectForKey:@"source"] intValue];
+    if ([_arguments objectForKey:@"maxDuration"]) {
+        double max = [[_arguments objectForKey:@"maxDuration"] doubleValue];
```
Comment (cyanglaz) [inline packages/image_picker/ios/Classes/ImagePickerPlugin.m]: nits: NSTimeInterval maxDuration

File: packages/image_picker/lib/image_picker.dart [hunk 2]
```diff
@@ -74,16 +74,18 @@ class ImagePicker {
   /// The [source] argument controls where the video comes from. This can
   /// be either [ImageSource.camera] or [ImageSource.gallery].
   ///
+  /// The [maxDuration] argument specifies the maximum duration of the captured video.
```
Comment (cyanglaz) [inline packages/image_picker/lib/image_picker.dart]: explain when [maxDuration] is null, the maxDuration is infinite

File: packages/image_picker/ios/Classes/ImagePickerPlugin.m (hunk 1 above)
Comment (cyanglaz) [reply packages/image_picker/ios/Classes/ImagePickerPlugin.m]: Also if maxDuration is NSNull from dart, this `max` here will be come 0.

**Comment by klaszlo8207** [comment]
Finally, this is awesome :)

File: packages/image_picker/lib/image_picker.dart (hunk 2 above)
Comment (derTuca) [reply packages/image_picker/lib/image_picker.dart]: Sorted

File: packages/image_picker/ios/Classes/ImagePickerPlugin.m (hunk 1 above)
Comment (derTuca) [reply packages/image_picker/ios/Classes/ImagePickerPlugin.m]: Could you explain a case where max would become 0 more clearly, please? Because if NSNull is passed from dart, the if shouldn't pass.

File: packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java [hunk 3]
```diff
@@ -249,7 +249,7 @@ public void chooseVideoFromGallery(MethodCall methodCall, MethodChannel.Result r
   }
 
   private void launchPickVideoFromGalleryIntent() {
-    Intent pickVideoIntent = new Intent(Intent.ACTION_GET_CONTENT);
+    Intent pickVideoIntent = new Intent(Intent.ACTION_PICK, MediaStore.Video.Media.EXTERNAL_CONTENT_URI);
```
Comment (cyanglaz) [inline packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java]: Could you revert this seems not related to this PR.

File: packages/image_picker/example/ios/Runner.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -340,16 +340,13 @@
...
-				"${PODS_ROOT}/../.symlinks/flutter/ios/Flutter.framework",
 			);
 			name = "[CP] Embed Pods Frameworks";
 			outputPaths = (
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Flutter.framework",
 			);
 			runOnlyForDeploymentPostprocessing = 0;
 			shellPath = /bin/sh;
-			shellScript = "\"${SRCROOT}/Pods/Target Support Files/Pods-Runner/Pods-Runner-frameworks.sh\"\n";
+			shellScript = "\"${PODS_ROOT}/Target Support Files/Pods-Runner/Pods-Runner-frameworks.sh\"\n";
```
Comment (cyanglaz) [inline packages/image_picker/example/ios/Runner.xcodeproj/project.pbxproj]: revert

File: packages/image_picker/example/ios/Runner.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -340,16 +340,13 @@
 			files = (
 			);
 			inputPaths = (
-				"${SRCROOT}/Pods/Target Support Files/Pods-Runner/Pods-Runner-frameworks.sh",
-				"${PODS_ROOT}/../.symlinks/flutter/ios/Flutter.framework",
 			);
 			name = "[CP] Embed Pods Frameworks";
 			outputPaths = (
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Flutter.framework",
```
Comment (cyanglaz) [inline packages/image_picker/example/ios/Runner.xcodeproj/project.pbxproj]: revert

File: packages/image_picker/example/ios/Runner.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -340,16 +340,13 @@
 			files = (
 			);
 			inputPaths = (
-				"${SRCROOT}/Pods/Target Support Files/Pods-Runner/Pods-Runner-frameworks.sh",
```
Comment (cyanglaz) [inline packages/image_picker/example/ios/Runner.xcodeproj/project.pbxproj]: revert

File: packages/image_picker/ios/Classes/ImagePickerPlugin.m (hunk 1 above)
Comment (cyanglaz) [reply packages/image_picker/ios/Classes/ImagePickerPlugin.m]: As far as I remember, 
`if ([_arguments objectForKey:@"maxDuration"])`  will return true if [_arguments objectForKey:@"maxDuration"] is NSNull.
You could also change it to 
`if ([[_arguments objectForKey:@"maxDuration"] isKindOf:[NSNumber class]])`

**Comment by klaszlo8207** [comment]
Hey, whats happening with this PR? Nothing?

**Comment by klaszlo8207** [comment]
Any news on this? Please help me

**Comment by cyanglaz** [comment]
@derTuca It is been a while since I reviewed this PR. Do you still want to continue to work on this?

**Comment by klaszlo8207** [comment]
@cyanglaz can you help me in this PR? It looks like @derTuca has no time for this. Can you fix it and merge to the base code?

Thanks!

**Comment by cyanglaz** [comment]
@derTuca Do you think you still want to work on this?

If anyone is interested to open another PR to work on this, I am happy to review it.

**Comment by klaszlo8207** [comment]
Can you fix the problemes in the conflicting files @cyanglaz? and then just simple attach to the base? It is a must to have feature, and I want this before last month :)

**Comment by derTuca** [comment]
Sorry for the huge delays, I am caught up with some stuff at this time. I have adopted the proposed changes, @cyanglaz if you could review it again it would be great.

**Comment by derTuca** [comment]
@cyanglaz should be fine now.

File: packages/image_picker/CHANGELOG.md [hunk 7]
```diff
@@ -1,344 +1,349 @@
+## 0.6.3
+
+- New feature : Set maximum duration for video recording.
```
Comment (cyanglaz) [inline packages/image_picker/CHANGELOG.md]: I believe this update of changing * to - was some mistake?

**Comment by ened** [comment]
Hi @derTuca could you kindly rebase to latest master and have another look at the CHANGELOG.md changes? The formatting for many entries is still a bit off.
The plugins repository is now building nicely again, so we can merge this PR soon. Thank you!

**Comment by cyanglaz** [comment]
@derTuca Please also add a jUnit test to test the max duration property in Android.

**Comment by derTuca** [comment]
@cyanglaz I haven't done much testing with Java, so I need some help with how to mock the ImagePickerDelegate class properly; I don't know if I just need to use @Mock or if I need to do a more...proper...initialization of the constructor's args.

**Comment by cyanglaz** [comment]
@derTuca There are some existing junit test cases in this plugin. You can take a look and see how they are constructed.

**Comment by klaszlo8207** [comment]
Hello all, any way to add this PR to the source code now?

**Comment by derTuca** [comment]
@cyanglaz Done, thanks for the info!

File: packages/image_picker/CHANGELOG.md [hunk 8]
```diff
@@ -78,8 +84,7 @@
 
 ## 0.6.1
 
-* New feature : Get images with custom quality. While picking images, user can pass `imageQuality`
-parameter to compress image.
+* New feature : Get images with custom quality. While picking images, user can pass `imageQuality` parameter to compress image.
```
Comment (cyanglaz) [inline packages/image_picker/CHANGELOG.md]: revert

**Comment by derTuca** [comment]
> Thanks for adding the java tests. We can also add iOS unit testing. There should be some unit test cases in this plugin as well.

Do you have any tips on how I can start on this? I have seen the other tests, but there are none testing passing a parameter to the controller per se (as far as I can tell), and I don't really know what to test here.

**Comment by cyanglaz** [comment]
@derTuca I would:
1. Create a `FLTImagePickerPlugin` with some mocking parameters. (Feel free to use OCMock). 
2. Pass maxDuration to the method channel.
3. Check if the maxDuration of the controller is the same as the one you passed in.

**Comment by derTuca** [comment]
For the life of everything, I swear that (with all the free time everybody suddenly has) I've looked into this, but I can't figure out how to check the duration in the controller. I can do the Android approach, where we just check if it returns the file path with the argument passed, but with the controller being private and all (so I can't mock it), I don't know how to check its maxDuration.

@cyanglaz am I missing the magic sauce? Or is it just straight up not possible to check it without modifying the plugin code?

**Comment by mehmetf** [comment]
FWIW, this is something customer:money cares about. @cyanglaz it would be great if you can help @derTuca here.

**Comment by cyanglaz** [comment]
> FWIW, this is something customer:money cares about. @cyanglaz it would be great if you can help @derTuca here.

Sure, I'll move this up in my priority list.

**Comment by cyanglaz** [comment]
@derTuca In my opinion, It is totally legit to modify the code in order to make them testable. Feel free to do so. Let me know if you still face troubles. :)

**Comment by cyanglaz** [comment]
@derTuca Just checking in how you are doing :) Are you able to continue working on this PR?

**Comment by ngoctranfire** [comment]
@derTuca, curious about your progress here too. This is something I'm looking into as well too.

**Comment by cyanglaz** [comment]
@derTuca I added the tests and resolved the conflicts for this PR in https://github.com/derTuca/plugins/pull/1
Do you mind taking a look and merge?

**Comment by derTuca** [comment]
> @derTuca I added the tests and resolved the conflicts for this PR in [derTuca#1](https://github.com/derTuca/plugins/pull/1)
> Do you mind taking a look and merge?

Merged, sorry for not being able to look into it after all.

**Comment by cyanglaz** [comment]
@derTuca It seems the pull request went to a wrong branch. Could you create another PR to the correct branch to fix that?

**Comment by derTuca** [comment]
Done. #2643.
